crates/router/src/core/payment_methods/transformers.rs

```diff
@@ -1,104 +1,220 @@
+}
+
+pub fn get_dotted_jws(jws: encryption::JwsBody) -> String {
+    let header = jws.header;
+    let payload = jws.payload;
+    let signature = jws.signature;
+    format!("{header}.{payload}.{signature}")
+}
+
+pub async fn get_decrypted_response_payload(
+    jwekey: &Jwekey,
+    jwe_body: encryption::JweBody,
+) -> CustomResult<String, errors::VaultError> {
+    let public_key = encryption::KeyHandler::get_kms_decrypted_key(
+        jwekey,
+        jwekey.vault_encryption_key.to_string(),
+    )
+    .await
+    .change_context(errors::VaultError::SaveCardFailed)
+    .attach_printable("Fails to get public key of vault")?;
+    let private_key =
+        encryption::KeyHandler::get_kms_decrypted_key(jwekey, jwekey.vault_private_key.to_string())
+            .await
+            .change_context(errors::VaultError::SaveCardFailed)
+            .attach_printable("Error getting private key for signing jws")?;
+
+    let jwt = get_dotted_jwe(jwe_body);
+    let key_id = basilisk_hs_key_id();
+    let jwe_decrypted = encryption::decrypt_jwe(jwekey, &jwt, key_id, key_id, private_key)
+        .await
+        .change_context(errors::VaultError::SaveCardFailed)
+        .attach_printable("Jwe Decryption failed for JweBody for vault")?;
+
+    let jws = jwe_decrypted
+        .parse_struct("JwsBody")
+        .change_context(errors::VaultError::ResponseDeserializationFailed)?;
+    let jws_body = get_dotted_jws(jws);
+
+    encryption::verify_sign(jws_body, &public_key)
+        .change_context(errors::VaultError::SaveCardFailed)
+        .attach_printable("Jws Decryption failed for JwsBody for vault")
+}
+
+pub async fn mk_basilisk_req(
+    jwekey: &Jwekey,
+    jws: &str,
+) -> CustomResult<encryption::JweBody, errors::VaultError> {
+    let jws_payload: Vec<&str> = jws.split('.').collect();
+    let jws_body = encryption::JwsBody {
+        header: jws_payload[0].to_owned(),
+        payload: jws_payload[1].to_owned(),
+        signature: jws_payload[2].to_owned(),
+    };
+    let payload = utils::Encode::<encryption::JwsBody>::encode_to_string_of_json(&jws_body)
+        .change_context(errors::VaultError::SaveCardFailed)?;
+
+    let public_key = encryption::KeyHandler::get_kms_decrypted_key(
+        jwekey,
+        jwekey.vault_encryption_key.to_string(),
+    )
+    .await
+    .change_context(errors::VaultError::SaveCardFailed)
+    .attach_printable("Fails to get encryption key of vault")?;
+
+    let jwe_encrypted = encryption::encrypt_jwe(jwekey, &payload, public_key)
+        .await
+        .change_context(errors::VaultError::SaveCardFailed)
+        .attach_printable("Error on jwe encrypt")?;
+    let jwe_payload: Vec<&str> = jwe_encrypted.split('.').collect();
+    Ok(encryption::JweBody {
+        header: jwe_payload[0].to_owned(),
+        iv: jwe_payload[2].to_owned(),
+        encrypted_payload: jwe_payload[3].to_owned(),
+        tag: jwe_payload[4].to_owned(),
+        encrypted_key: jwe_payload[1].to_owned(),
```

Comment:
If there is any issue, while splitting this, the code could panic. Can we use some safer functionality like `.get()` or have a implementation of `FromStr` something like `impl FromStr for JweBody`?

Reply:
added check on length